### ISSUE TITLE
ConfigRepo pipeline does not respect default job timeout

### ISSUE THREAD

**Issue by sandro-h** [gocd/gocd#6302]
##### Issue Type

- Bug Report

##### Summary

Since upgrading from GoCD 18.11 to 19.2, it appears stuck pipelines are not automatically cancelled anymore.
We set the default server timeout to 90 minutes, but had warnings like this:
```
Job xyz/build/build is currently running but has not shown any console activity in the last 1485 minute(s). This job may be hung.
```

We use the [gocd-yaml-config-plugin](https://github.com/tomzo/gocd-yaml-config-plugin) to configure all our pipelines.
I think this may be a regression in the handling of default timeouts for config repo pipelines since 19.1. See at the bottom.

##### Environment

GoCD server 19.2.0
YAML Configuration Plugin 0.9.0

##### Steps to Reproduce

1.  Set up a pipeline YAML config without any explicit job timeouts set and let it wait for 10 minutes. E.g. with something like:
```
jobs:
  wait:
    tasks:
      - exec:
         command: /bin/sh
           arguments: ["-c", "sleep 600"]
```
2. Set the server default timeout to 5 minutes
3. Run the pipeline

##### Expected Results

After 5 minutes, the wait job is cancelled.

##### Actual Results

Nothing happens, no cancel attempt is made.

##### Any other info

I suspect this change may be the culprit: https://github.com/gocd/gocd/commit/680980e7789d604a19a8428c7885c66b0602d58a

Before this, when the config repo plugin sent no explicit timeout, the timeout would default to `null` and the server default timeout would be used. Now it defaults to 0, i.e. "never".

This can also be seen when querying the pipeline config in the API:
```
curl "https://my-gocd-server.com:8154/go/api/admin/pipelines/xyz" -s \
  -u $USER -H 'Accept: application/vnd.go.cd.v6+json'
```
```
[...]
    "jobs" : [ {
      "name" : "wait",
      "run_instance_count" : null,
      "timeout" : "never",
[...]
```

When doing the same with a pipeline configured in the GUI, the API correctly returns `"timeout" : null`.

**Comment by ankitsri11** [comment]
@sandro-h Thank you for reporting the issue. I was able to replicate it locally.

As mentioned by you, I tried with below yaml config-repo pipeline and set the server level time out to 5minutes. 

After 5 minutes, the pipeline was still running and dint honor the server time out configuration.

<img width="696" alt="Screenshot 2019-05-17 at 11 02 32" src="https://user-images.githubusercontent.com/6838313/57947286-569a7700-7893-11e9-9715-c9963b0b1e48.png">

```
format_version: 3
pipelines:
  PipelineA:
    group: defaultGroup
    label_template: ${COUNT}
    lock_behavior: none
    display_order: -1
    materials:
      git:
        git: https://github.com/some-repo
        shallow_clone: false
        auto_update: true
        branch: master
    stages:
    - defaultStage:
        fetch_materials: true
        keep_artifacts: false
        clean_workspace: false
        approval:
          type: success
        jobs:
          defaultJob:
            timeout: 0
            tasks:
            - exec:
                arguments:
                - -c
                - sleep 600
                command: /bin/bash
                run_if: passed 
```

API output:

```
curl 'http://localhost:8153/go/api/admin/pipelines/PipelineA' \
      -u 'username:password' \
      -H 'Accept: application/vnd.go.cd.v6+json' \
      -i
```

```
  "jobs": [
                {
                    "name": "defaultJob",
                    "run_instance_count": null,
                    "timeout": "never",
                    "environment_variables": [],
                    "resources": [],
                    "tasks": [
                        {
                            "type": "exec",
                            "attributes": {
                                "run_if": [
                                    "passed"
                                ],
                                "command": "/bin/bash",
                                "arguments": [
                                    "-c",
                                    "sleep 600"
                                ]
                            }
                        }
                    ],
```

**Comment by ankitsri11** [comment]
@sandro-h we have fixed it as part of this commit https://github.com/gocd/gocd/commit/48c9abb68f9cabaab2382c901268a1b123129c24. It should be available in the next GoCD release 19.4.0. 

If you want to try it out and share feedback, you can use the experimental build listed on the GoCD download page: https://www.gocd.org/download/ 

I wouldn't recommend using it in the production environment. Thanks.

**Comment by sandro-h** [comment]
@ankitsri11 Ah, great :) Thanks for the quick response!

I tried the experimental build (19.4.0-9113) and it cancelled the build after 5 minutes as expected.
